### ISSUE TITLE
Catch usages of sun.misc.Unsafe or jdk.internal.misc.Unsafe

### ISSUE THREAD

**Issue by thomasleplus** [pmd/pmd#5923]
**Proposed Rule Name:**

**Proposed Category:** Security

**Description:**

**Code Sample:** This should include code, that should be flagged by the rule. If possible, the "correct" code
according to this new rule should also be demonstrated.

In the bad code below, 128 bytes of memory are allocated but then arbitrary strings are written which could easily go over the 128 bytes limit.

```java
import java.lang.reflect.Field;
import sun.misc.Unsafe;

public final class UnsafePutChar {

  public static void main(final String[] args) throws NoSuchFieldException, IllegalAccessException {
    final long address = getUnsafe().allocateMemory(128);
    for (final String s : args) {
      for (final char c : s.toCharArray()) {
        getUnsafe().putChar(address, c);
      }
    }
  }

  private static Unsafe getUnsafe() throws IllegalAccessException, NoSuchFieldException {
    final Field f = Unsafe.class.getDeclaredField("theUnsafe");
    f.setAccessible(true);
    return (Unsafe) f.get(null);
  }
}
```

The right approach if using off-heap memory is really necessary would be to use the [Foreign Function and Memory API](https://docs.oracle.com/en/java/javase/21/core/foreign-function-and-memory-api.html).

In general I think that any use of `sun.misc.Unsafe` or `jdk.internal.misc.Unsafe` should be flagged as a security concern. I know that the `DontImportSun` already covers `sun.misc.Unsafe` but it is wider in its scope (all of `sun.*`) and it doesn't explain the security risk so it might be easily dismissed by someone who doesn't know the particular security implications of `sun.misc.Unsafe`. A security-specific rule including both `sun.misc.Unsafe` and `jdk.internal.misc.Unsafe` seems more appropriate.

I can create a PR for this but I wanted to discuss it first. See a working draft of XPath rule [here](https://github.com/thomasleplus/java-memory-safety/blob/main/pmd/unsafe-ruleset.xml).

**Possible Properties:**

*   Should this rule be customizable via properties?

Not that I can think of.

**Comment by elharo** [comment]
JDK already warns about this. I'm not sure why PMD should duplicate that.
